### ISSUE TITLE
0.96.0 beta 5 creating new script forces exit upon renaming or setting a hotkey

### ISSUE THREAD

**Issue by kreezxil** [autokey/autokey#564]
## Classification:

crash

## Reproducibility:

always

## Version

AutoKey version: 0.96.0 beta 5

Used GUI (Gtk, Qt, or both): Qt

If the problem is known to be present in more than one version, please list all of those.

Installed via: (PPA, pip3, …). pip

Linux Distribution: Manjaro KDE Plasma

## Full System Specs
```
Operating System: Manjaro Linux
KDE Plasma Version: 5.21.5
KDE Frameworks Version: 5.82.0
Qt Version: 5.15.2
Kernel Version: 5.12.9-1-MANJARO
OS Type: 64-bit
Graphics Platform: X11
Processors: 8 × AMD FX(tm)-8350 Eight-Core Processor
Memory: 23.5 GiB of RAM
Graphics Processor: NVIDIA GeForce GTX 750 Ti/PCIe/SSE2
```

## Summary

Creating a script and then renaming it or setting a hotkey causes it to crash out.

## log from --verbose
https://gist.github.com/f4d7a4b1e298c663301ef6c2d5721e0a

## Workaround
Falling back to 0.96.0 beta 4 resolves this issue.

### use this for global install:
```
sudo pip3 install https://github.com/autokey/autokey/archive/v0.96.0-beta.4.zip
```
### use this for local install:
```
pip3 install --user https://github.com/autokey/autokey/archive/v0.96.0-beta.4.zip
```
#### notes before installing another version
I fully recommend removing the other other versions, if you used `pip` you can do
```
pip3 uninstall autokey
```
**you may** need to run it a few times, if you have more than one autokey installed.

**Comment by josephj11** [comment]
One small clarification: as we use the terms, a script is a Python script that AutoKey runs when it is invoked. A macro is a construct used within a phrase that adds additional functionality. It is specified by surrounding it in the `<>` tokens. There's aa section in the wiki called Dynamic Phrases, but the link is currently broken.

**Comment by kreezxil** [comment]
What does your clarification have to do with my bug report? I'll make a video since the words are lost to you.

**Comment by kreezxil** [comment]
https://user-images.githubusercontent.com/3880072/122678044-a2265a00-d1aa-11eb-8fa5-b80a5097a529.mp4

**Comment by josephj11** [comment]
> What does your clarification have to do with my bug report? I'll make a video since the words are lost to you.

Your issue report is fine and I understood it. That's why I gave it the appropriate labels.

I am addressing the wording of your title. In AutoKey, phrases can have macros. Within AutoKey, scripts are scripts. There is no such thing as a script macro unless, perhaps, you are using a macro in a phrase to invoke a script.

**Comment by kreezxil** [comment]
You see the video now. You are welcome to rename it how you see fit. Sorry if I came off as on edge.

**Comment by josephj11** [comment]
NP. I've been known to get snippy upon occasion. :)

Just wondering: Does renaming an "old" script work fine?

When you subsequently restart AutoKey, is the script still named "New Script"?

Does anything show up in the logs or in an AutoKey trace as to what AutoKey is upset about?

Could this have anything to do with the fact that you have three scripts named "New Script" - which invokes code in AutoKey to make the file names unique under the hood? To test this, you  could first close AutoKey and manually rename the other two scripts and their sidecar files in your file manager or terminal.

**Comment by kreezxil** [comment]
it has nothing to do with those. those are artifacts from trying to even
create a script.

**Comment by josephj11** [comment]
I assumed that those were caused by not being able to rename them, but can you rename an old script?

**Comment by kreezxil** [comment]
i'll give it a go.

**Comment by kreezxil** [comment]
Still crashes. Here's a video.

https://user-images.githubusercontent.com/3880072/123826492-6e8fb200-d8c5-11eb-95a2-5e10429c23c3.mp4

**Comment by josephj11** [comment]
Would it now be correct to say that renaming any script is the issue rather than just new ones?

Is this true with setting a new hotkey for any script too?

**Comment by ineuw** [comment]
My first approach as an ignoramus, would be to check out the contents of the script folders. I often discovered problems because of typos, or unmatched files.

**Comment by kreezxil** [comment]
Is no one else having issues making new scripts with beta 5? Why would it
work fine on beta 4 and lower and not on beta 5? Why does have to be
associated with ignorance? Or rather why can't the program say "Oh, this is
old code" and then popup something that says "you're going to get problems
with this, I recommend doing a 'reset'" and there is a box there, and you
hit the option to do that reset and the old configs are wiped, and new ones
are generated?

I rely on this macro/script triggering program heavily, I want it to work
flawlessly and not treat me like it's an elitist Debian or Arch developer.

**Comment by josephj11** [comment]
@kreezxil You sound a bit upset. AFAIK, this looks like a normal beta bug. It needs to be fixed, but it's no big deal as long as it doesn't make it into a release. The point of having a beta is to find things like this.

I don't know how many people are actually actively using the beta. It may be very few  which would explain why you haven't seen any "me too" responses yet.

I haven't installed it on my system yet. But I did do some prep work for it and modified my system so that it will switch between the old and new versions by using separate config trees and a symlink to the one in use. This way, I can go back to 0.95.10 at any time to get work done if 0.96 is misbehaving.

**Comment by kreezxil** [comment]
I'm not upset. It's just that I report the issues, and then I feel I'm
getting responses in the ticket from folks telling me "it ain't so". Now I
wonder if they're even trying to duplicate it on their end with this
particular issue before sounding off. At least I can freely move between
versions now. I didn't know how to do that before.

**Comment by josephj11** [comment]
At least in this issue, no one has commented that they couldn't recreate this (or that they have even tried to do so.)

It is quite typical for issues to affect different systems in different ways - especially in Linux where individual systems can be quite different. If a bug affects one system, but not another, it can help narrow down what the problem is - especially when it relates to dependencies or the configurations themselves.

**Comment by kreezxil** [comment]
ok, i'll take it upon myself to see if I can recreate this. I put my specs up top already, so you know the current setup where the issue is showing. I'll now try the latest Ubuntu in my VM.

**Comment by kreezxil** [comment]
I have duplicated the issue on Ubuntu 21.04 Hirsute (Wayland), had to install kdialog to make it work. So KDE seems to be a common thread then. I'll try gnome next.

![image](https://user-images.githubusercontent.com/3880072/124390774-ac6c4c00-dcb2-11eb-8c05-ed62d50cba0a.png)


https://user-images.githubusercontent.com/3880072/124390806-cdcd3800-dcb2-11eb-81de-3e79b781b1ff.mp4

**Comment by kreezxil** [comment]
I added gnome to the vm, but it's still on wayland. Will reinstall from a livecd now to ensure a proper gnome Ubuntu environment. It's been so long since I used Ubuntu last. I don't see an actual Gnome distribution for it. I see a hybrid called Kylin tho. I'll try that. I saw MATE mentioned, so I'll try that afterwards.

**Comment by kreezxil** [comment]
Interesting, this is probably another issue altogether, and it's made evident by new users or new installations. You existing Ubuntu users have artifacts from your previous versions that allow Autokey betas to run right out of the box.

Here's some of what I have to go through just to get the beta to install on each fresh system "Ubuntu" as above, now "Mate".

```bash
kreezxil@kreezxil-VM-MATE:~/Desktop$ autokey-qt
Traceback (most recent call last):
  File "/usr/local/bin/autokey-qt", line 5, in <module>
    from autokey.qtui.__main__ import Application
  File "/usr/local/lib/python3.9/dist-packages/autokey/qtui/__main__.py", line 22, in <module>
    from PyQt5 import QtCore
ModuleNotFoundError: No module named 'PyQt5'
kreezxil@kreezxil-VM-MATE:~/Desktop$ sudo pip install pyqt5
Collecting pyqt5
  Downloading PyQt5-5.15.4-cp36.cp37.cp38.cp39-abi3-manylinux2014_x86_64.whl (8.3 MB)
     |████████████████████████████████| 8.3 MB 4.1 MB/s 
Collecting PyQt5-Qt5>=5.15
  Downloading PyQt5_Qt5-5.15.2-py3-none-manylinux2014_x86_64.whl (59.9 MB)
     |████████████████████████████████| 59.9 MB 17.3 MB/s 
Collecting PyQt5-sip<13,>=12.8
  Downloading PyQt5_sip-12.9.0-cp39-cp39-manylinux1_x86_64.whl (328 kB)
     |████████████████████████████████| 328 kB 14.3 MB/s 
Installing collected packages: PyQt5-sip, PyQt5-Qt5, pyqt5
Successfully installed PyQt5-Qt5-5.15.2 PyQt5-sip-12.9.0 pyqt5-5.15.4
kreezxil@kreezxil-VM-MATE:~/Desktop$ autokey-qt
Traceback (most recent call last):
  File "/usr/local/bin/autokey-qt", line 5, in <module>
    from autokey.qtui.__main__ import Application
  File "/usr/local/lib/python3.9/dist-packages/autokey/qtui/__main__.py", line 23, in <module>
    from autokey.qtapp import Application
  File "/usr/local/lib/python3.9/dist-packages/autokey/qtapp.py", line 44, in <module>
    from autokey.qtui.configwindow import ConfigWindow
  File "/usr/local/lib/python3.9/dist-packages/autokey/qtui/configwindow.py", line 40, in <module>
    class ConfigWindow(*autokey.qtui.common.inherits_from_ui_file_with_name("mainwindow")):
  File "/usr/local/lib/python3.9/dist-packages/autokey/qtui/common.py", line 145, in load_ui_from_file
    base_type = uic.loadUiType(ui_file, from_imports=True)
  File "/usr/local/lib/python3.9/dist-packages/PyQt5/uic/__init__.py", line 204, in loadUiType
    exec(code_string.getvalue(), ui_globals)
  File "<string>", line 293, in <module>
  File "/usr/local/lib/python3.9/dist-packages/autokey/qtui/centralwidget.py", line 42, in <module>
    class CentralWidget(*ui_common.inherits_from_ui_file_with_name("centralwidget")):
  File "/usr/local/lib/python3.9/dist-packages/autokey/qtui/common.py", line 145, in load_ui_from_file
    base_type = uic.loadUiType(ui_file, from_imports=True)
  File "/usr/local/lib/python3.9/dist-packages/PyQt5/uic/__init__.py", line 204, in loadUiType
    exec(code_string.getvalue(), ui_globals)
  File "<string>", line 67, in <module>
  File "/usr/local/lib/python3.9/dist-packages/autokey/qtui/scriptpage.py", line 20, in <module>
    from PyQt5 import Qsci
ImportError: cannot import name 'Qsci' from 'PyQt5' (/usr/local/lib/python3.9/dist-packages/PyQt5/__init__.py)
kreezxil@kreezxil-VM-MATE:~/Desktop$ sudo pip install qsci
ERROR: Could not find a version that satisfies the requirement qsci
ERROR: No matching distribution found for qsci
kreezxil@kreezxil-VM-MATE:~/Desktop$ autokey-gtk
Traceback (most recent call last):
  File "/usr/local/bin/autokey-gtk", line 5, in <module>
    from autokey.gtkui.__main__ import main
  File "/usr/local/lib/python3.9/dist-packages/autokey/gtkui/__main__.py", line 4, in <module>
    from autokey.gtkapp import Application
  File "/usr/local/lib/python3.9/dist-packages/autokey/gtkapp.py", line 40, in <module>
    from autokey.gtkui.notifier import get_notifier
  File "/usr/local/lib/python3.9/dist-packages/autokey/gtkui/notifier.py", line 26, in <module>
    gi.require_version('AppIndicator3', '0.1')
  File "/usr/lib/python3/dist-packages/gi/__init__.py", line 126, in require_version
    raise ValueError('Namespace %s not available' % namespace)
ValueError: Namespace AppIndicator3 not available
kreezxil@kreezxil-VM-MATE:~/Desktop$ 
```
source: https://hastebin.com/maqobapogi.sql

**Comment by kreezxil** [comment]
Ok, I figured out that Qsci is Qscintilla. Is there no way to make the following command more intelligent with pip, such that installs the dependencies? For Linux Mate there is only -GTK and -qt, installing the -qt variant causes this to happen.

![Image](https://user-images.githubusercontent.com/3880072/124392934-a465d980-dcbd-11eb-9bb4-1624110b4ee6.png)

So, I thought I would go the -GTK route specifically this time, and immediately I've run into issues. Googling says it's part of the PyGObject package which is already installed, but Autokey says, "uh, no it isn't". So, then I discovered you can explicitly install Python GTK by using `sudo pip install pygtk` but then I'm met with a message that says "autobuilding" isn't directly supported except on Windows.

```bash
kreezxil@kreezxil-VM-MATE:~/Desktop$ sudo pip install pygtk
Collecting pygtk
  Downloading pygtk-2.24.0.tar.bz2 (2.4 MB)
     |████████████████████████████████| 2.4 MB 5.3 MB/s 
    ERROR: Command errored out with exit status 1:
     command: /usr/bin/python3 -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-5753b58j/pygtk_8563be33d30b45098f9e5f09679ad58e/setup.py'"'"'; __file__='"'"'/tmp/pip-install-5753b58j/pygtk_8563be33d30b45098f9e5f09679ad58e/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-pip-egg-info-b1gmv0vs
         cwd: /tmp/pip-install-5753b58j/pygtk_8563be33d30b45098f9e5f09679ad58e/
    Complete output (4 lines):
    ********************************************************************
    * Building PyGTK using distutils is only supported on windows. *
    * To build PyGTK in a supported way, read the INSTALL file.    *
    ********************************************************************
    ----------------------------------------
WARNING: Discarding https://files.pythonhosted.org/packages/7c/18/fa4f2de77500dd62a314fd845ff6e903ac2ce551164cb421c5750969f799/pygtk-2.24.0.tar.bz2#sha256=cd1c1ea265bd63ff669e92a2d3c2a88eb26bcd9e5363e0f82c896e649f206912 (from https://pypi.org/simple/pygtk/). Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
ERROR: Could not find a version that satisfies the requirement pygtk
ERROR: No matching distribution found for pygtk
kreezxil@kreezxil-VM-MATE:~/Desktop$ 
``` Source: https://hastebin.com/refazebuda.sql

Unless someone can tell me how to satisfy that for now, I'm forced to go down the kdialog route, and you know what I'll be reporting next because I think it's related to kdialog.

**I fully recommend** that all developers on this app install fresh copies of their operating systems in VMs, and then right away install Autokey, note any extra steps you're taking and add that information to the wiki and then work on getting that into the installer. 

Well MATE is a no-go. so sad. 
![image](https://user-images.githubusercontent.com/3880072/124396302-4c839e80-dcce-11eb-800c-b55c096fe4f9.png)

```bash
kreezxil@kreezxil-VM-MATE:~/Desktop$ autokey-qt
2021-07-04 13:44:37,932 INFO - autokey.UI_common_functions - Commandline program "visgrep" was not found/able to be used correctly by AutoKey. This program is optional for Autokey operation, but if you wish to use functionality associated with it, Check that "visgrep" exists on your system and is in the $PATH seen by Autokey.
2021-07-04 13:44:37,932 INFO - autokey.UI_common_functions - Commandline program "import" was not found/able to be used correctly by AutoKey. This program is optional for Autokey operation, but if you wish to use functionality associated with it, Check that "import" exists on your system and is in the $PATH seen by Autokey.
2021-07-04 13:44:37,932 INFO - autokey.UI_common_functions - Commandline program "png2pat" was not found/able to be used correctly by AutoKey. This program is optional for Autokey operation, but if you wish to use functionality associated with it, Check that "png2pat" exists on your system and is in the $PATH seen by Autokey.
2021-07-04 13:44:37,932 INFO - autokey.UI_common_functions - Commandline program "xte" was not found/able to be used correctly by AutoKey. This program is optional for Autokey operation, but if you wish to use functionality associated with it, Check that "xte" exists on your system and is in the $PATH seen by Autokey.
2021-07-04 13:44:37,933 INFO - autokey.UI_common_functions - Commandline program "xmousepos" was not found/able to be used correctly by AutoKey. This program is optional for Autokey operation, but if you wish to use functionality associated with it, Check that "xmousepos" exists on your system and is in the $PATH seen by Autokey.
2021-07-04 13:44:37,933 INFO - autokey.qtapp - Initialising application
2021-07-04 13:44:38,016 INFO - autokey.qtapp - Initialise global hotkeys
2021-07-04 13:44:38,018 INFO - autokey.configmanager.configmanager - No configuration found - creating new one
2021-07-04 13:44:38,018 INFO - autokey.configmanager.predefined_user_files - About to create folder 'My Phrases'
2021-07-04 13:44:38,019 INFO - autokey.configmanager.predefined_user_files - About to create folder 'Addresses'
2021-07-04 13:44:38,023 INFO - autokey.configmanager.predefined_user_files - About to create folder 'Sample Scripts'
2021-07-04 13:44:38,024 ERROR - autokey.configmanager.configmanager - Error while loading configuration. Cannot recover.
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/autokey/configmanager/configmanager.py", line 53, in create_config_manager_instance
    config_manager = ConfigManager(auto_key_app)
  File "/usr/local/lib/python3.9/dist-packages/autokey/configmanager/configmanager.py", line 229, in __init__
    self.folders.append(autokey.configmanager.predefined_user_files.create_sample_scripts_folder())
  File "/usr/local/lib/python3.9/dist-packages/autokey/configmanager/predefined_user_files.py", line 218, in create_sample_scripts_folder
    _create_script(item_data, sample_scripts)
  File "/usr/local/lib/python3.9/dist-packages/autokey/configmanager/predefined_user_files.py", line 147, in _create_script
    with open(str(content_path), "r", encoding="utf-8") as source_file:
FileNotFoundError: [Errno 2] No such file or directory: '/usr/local/lib/python3.9/dist-packages/autokey/configmanager/predefined_user_scripts/insert_date.pyi'
2021-07-04 13:44:38,026 ERROR - autokey.qtapp - Fatal error starting AutoKey: [Errno 2] No such file or directory: '/usr/local/lib/python3.9/dist-packages/autokey/configmanager/predefined_user_scripts/insert_date.pyi'
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/autokey/qtapp.py", line 122, in __init__
    self.configManager = cm.create_config_manager_instance(self)
  File "/usr/local/lib/python3.9/dist-packages/autokey/configmanager/configmanager.py", line 53, in create_config_manager_instance
    config_manager = ConfigManager(auto_key_app)
  File "/usr/local/lib/python3.9/dist-packages/autokey/configmanager/configmanager.py", line 229, in __init__
    self.folders.append(autokey.configmanager.predefined_user_files.create_sample_scripts_folder())
  File "/usr/local/lib/python3.9/dist-packages/autokey/configmanager/predefined_user_files.py", line 218, in create_sample_scripts_folder
    _create_script(item_data, sample_scripts)
  File "/usr/local/lib/python3.9/dist-packages/autokey/configmanager/predefined_user_files.py", line 147, in _create_script
    with open(str(content_path), "r", encoding="utf-8") as source_file:
FileNotFoundError: [Errno 2] No such file or directory: '/usr/local/lib/python3.9/dist-packages/autokey/configmanager/predefined_user_scripts/insert_date.pyi'
kreezxil@kreezxil-VM-MATE:~/Desktop$ sudo pip3 uninstall autokey==0.96.0-beta-5
[sudo] password for kreezxil: 
Found existing installation: autokey 0.96.0b5
Uninstalling autokey-0.96.0b5:
  Would remove:
    /usr/local/bin/autokey-gtk
    /usr/local/bin/autokey-qt
    /usr/local/bin/autokey-run
    /usr/local/bin/autokey-shell
    /usr/local/lib/python3.9/dist-packages/autokey-0.96.0b5.dist-info/*
    /usr/local/lib/python3.9/dist-packages/autokey/*
    /usr/local/share/applications/autokey-gtk.desktop
    /usr/local/share/applications/autokey-qt.desktop
    /usr/local/share/icons/Humanity/scalable/apps/autokey-status-error.svg
    /usr/local/share/icons/Humanity/scalable/apps/autokey-status.svg
    /usr/local/share/icons/hicolor/96x96/apps/autokey.png
    /usr/local/share/icons/hicolor/scalable/apps/autokey-status-dark.svg
    /usr/local/share/icons/hicolor/scalable/apps/autokey-status-error.svg
    /usr/local/share/icons/hicolor/scalable/apps/autokey-status.svg
    /usr/local/share/icons/hicolor/scalable/apps/autokey.svg
    /usr/local/share/icons/ubuntu-mono-dark/apps/48/autokey-status-error.svg
    /usr/local/share/icons/ubuntu-mono-dark/apps/48/autokey-status.svg
    /usr/local/share/icons/ubuntu-mono-light/apps/48/autokey-status-error.svg
    /usr/local/share/icons/ubuntu-mono-light/apps/48/autokey-status.svg
    /usr/local/share/man/man1/autokey-gtk.1
    /usr/local/share/man/man1/autokey-qt.1
    /usr/local/share/man/man1/autokey-run.1
Proceed (y/n)? y
  Successfully uninstalled autokey-0.96.0b5
kreezxil@kreezxil-VM-MATE:~/Desktop$ sudo pip3 install autokey==0.96.0-beta-5
WARNING: Keyring is skipped due to an exception: Failed to open keyring: org.freedesktop.DBus.Error.NoReply: Did not receive a reply. Possible causes include: the remote application did not send a reply, the message bus security policy blocked the reply, the reply timeout expired, or the network connection was broken..
Collecting autokey==0.96.0-beta-5
  Using cached autokey-0.96.0b5-py3-none-any.whl (395 kB)
Requirement already satisfied: packaging in /usr/local/lib/python3.9/dist-packages (from autokey==0.96.0-beta-5) (21.0)
Requirement already satisfied: python-xlib in /usr/lib/python3/dist-packages (from autokey==0.96.0-beta-5) (0.29)
Requirement already satisfied: pyinotify in /usr/local/lib/python3.9/dist-packages (from autokey==0.96.0-beta-5) (0.9.6)
Requirement already satisfied: pyparsing>=2.0.2 in /usr/local/lib/python3.9/dist-packages (from packaging->autokey==0.96.0-beta-5) (2.4.7)
Installing collected packages: autokey
Successfully installed autokey-0.96.0b5
kreezxil@kreezxil-VM-MATE:~/Desktop$ autokey-qt
2021-07-04 13:46:23,807 INFO - autokey.UI_common_functions - Commandline program "visgrep" was not found/able to be used correctly by AutoKey. This program is optional for Autokey operation, but if you wish to use functionality associated with it, Check that "visgrep" exists on your system and is in the $PATH seen by Autokey.
2021-07-04 13:46:23,808 INFO - autokey.UI_common_functions - Commandline program "import" was not found/able to be used correctly by AutoKey. This program is optional for Autokey operation, but if you wish to use functionality associated with it, Check that "import" exists on your system and is in the $PATH seen by Autokey.
2021-07-04 13:46:23,808 INFO - autokey.UI_common_functions - Commandline program "png2pat" was not found/able to be used correctly by AutoKey. This program is optional for Autokey operation, but if you wish to use functionality associated with it, Check that "png2pat" exists on your system and is in the $PATH seen by Autokey.
2021-07-04 13:46:23,808 INFO - autokey.UI_common_functions - Commandline program "xte" was not found/able to be used correctly by AutoKey. This program is optional for Autokey operation, but if you wish to use functionality associated with it, Check that "xte" exists on your system and is in the $PATH seen by Autokey.
2021-07-04 13:46:23,808 INFO - autokey.UI_common_functions - Commandline program "xmousepos" was not found/able to be used correctly by AutoKey. This program is optional for Autokey operation, but if you wish to use functionality associated with it, Check that "xmousepos" exists on your system and is in the $PATH seen by Autokey.
2021-07-04 13:46:23,809 INFO - autokey.qtapp - Initialising application
2021-07-04 13:46:23,892 INFO - autokey.qtapp - Initialise global hotkeys
2021-07-04 13:46:23,893 INFO - autokey.configmanager.configmanager - No configuration found - creating new one
2021-07-04 13:46:23,893 INFO - autokey.configmanager.predefined_user_files - About to create folder 'My Phrases'
2021-07-04 13:46:23,894 INFO - autokey.configmanager.predefined_user_files - About to create folder 'Addresses'
2021-07-04 13:46:23,896 INFO - autokey.configmanager.predefined_user_files - About to create folder 'Sample Scripts'
2021-07-04 13:46:23,896 ERROR - autokey.configmanager.configmanager - Error while loading configuration. Cannot recover.
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/autokey/configmanager/configmanager.py", line 53, in create_config_manager_instance
    config_manager = ConfigManager(auto_key_app)
  File "/usr/local/lib/python3.9/dist-packages/autokey/configmanager/configmanager.py", line 229, in __init__
    self.folders.append(autokey.configmanager.predefined_user_files.create_sample_scripts_folder())
  File "/usr/local/lib/python3.9/dist-packages/autokey/configmanager/predefined_user_files.py", line 218, in create_sample_scripts_folder
    _create_script(item_data, sample_scripts)
  File "/usr/local/lib/python3.9/dist-packages/autokey/configmanager/predefined_user_files.py", line 147, in _create_script
    with open(str(content_path), "r", encoding="utf-8") as source_file:
FileNotFoundError: [Errno 2] No such file or directory: '/usr/local/lib/python3.9/dist-packages/autokey/configmanager/predefined_user_scripts/insert_date.pyi'
2021-07-04 13:46:23,898 ERROR - autokey.qtapp - Fatal error starting AutoKey: [Errno 2] No such file or directory: '/usr/local/lib/python3.9/dist-packages/autokey/configmanager/predefined_user_scripts/insert_date.pyi'
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/autokey/qtapp.py", line 122, in __init__
    self.configManager = cm.create_config_manager_instance(self)
  File "/usr/local/lib/python3.9/dist-packages/autokey/configmanager/configmanager.py", line 53, in create_config_manager_instance
    config_manager = ConfigManager(auto_key_app)
  File "/usr/local/lib/python3.9/dist-packages/autokey/configmanager/configmanager.py", line 229, in __init__
    self.folders.append(autokey.configmanager.predefined_user_files.create_sample_scripts_folder())
  File "/usr/local/lib/python3.9/dist-packages/autokey/configmanager/predefined_user_files.py", line 218, in create_sample_scripts_folder
    _create_script(item_data, sample_scripts)
  File "/usr/local/lib/python3.9/dist-packages/autokey/configmanager/predefined_user_files.py", line 147, in _create_script
    with open(str(content_path), "r", encoding="utf-8") as source_file:
FileNotFoundError: [Errno 2] No such file or directory: '/usr/local/lib/python3.9/dist-packages/autokey/configmanager/predefined_user_scripts/insert_date.pyi'
```
source: https://hastebin.com/jukirocozo.sql

it complains heavily about missing scripts.

**Comment by kreezxil** [comment]
I'm going to call Mate a bust, I'll need someone on Mate to help me figure this out so we can get around to testing Autokey there.

**Comment by kreezxil** [comment]
Ok, this sucks, GTK does come preinstalled in kylin, but Autokey won't use it.
![image](https://user-images.githubusercontent.com/3880072/124397803-b3f21c00-dcd7-11eb-89df-d8da5af01038.png)

Kylin has the same issue as the above Mate issue. 
![image](https://user-images.githubusercontent.com/3880072/124397862-11866880-dcd8-11eb-8c3c-fc6284957e53.png)

The only Ubuntu I could get to work was the stock version running Wayland. And only after installing **Qscintilla** and **kdialog**, rename results as above.

Please before anyone adds comments to this, I urge, no, I beg you, to install this beta on a fresh copy of your OS, as in a VM, and try to duplicate this.

If you do any steps outside of 

```md
1. install pip # however your os handles that
2. pip install qscintilla
3. pip install autokey==0.96.0-beta-5
# should be noted i used sudo on everything
```

Then, please list those steps.

**Comment by josephj11** [comment]
Thanks for the work you've done.

Just a couple of observations - don't know how relevant they are.

We just had an issue with the new version of Mint/Mate where one piece of software is no longer compatible with us. I don't remember the details, but it's probably in an issue here.

AutoKey doesn't work under Wayland. It needs real X11. Wayland has some compatibility stuff for X11 apps, but not enough. It might be that their security design model makes it impossible for AutoKey to properly talk to the app windows. We have looked into this a bit, but there are no final conclusions.

I notice a message about 122 packages not upgraded. This sounds like you did a fresh install, but didn't update/upgrade before trying AutoKey. It probably should work, but it's better to have all the updates that were made after the release was built.

if you're clever, you may be able to install all that stuff once (before doing anything with AutoKey) and save that VM to use as a fresh starting point each time you want to install or test anything.

**Comment by kreezxil** [comment]
i thought that testing from a stock distro would be helpful. so i forewent
the time constraint involved in upgrading the distros.

**Comment by BlueDrink9** [comment]
Thank you very much for putting in this repro work kreezil, it is really appreciated. I'll take a look at this bug, although I can't guarantee fixing it soon.

Regarding reproducible installs, we definitely have issues. Some if it is to do with having two independent different sets of dependencies for qt vs gtk, and not wanting to install both of them for a user that only wants one set. I'm working in fixing that.
Another issue is that some of our dependencies can break the system if pip installs them rather than being installed via package manager.

I have set up the CI so that a fresh install is started and briefly run, to at least ensure non-gui components run correctly on a fresh install. But there is only so much that can be done on CI, and running the actual GUI is not one of them. I'm still surprised at how many issues you are having, so I'll see what can be done. To be honest, reproducing fresh installs each time is a bit of a nuisance and an effort, and I'm just not super keen to set myself up to do so outside of the CIs. However, the work you have done and the traces provided is awesome, and it'd be a great help to the community if you were willing to do the work you have done here for more versions!

Finally, regarding feeling like you are being asked to do all the reproduction work: that's a valid feeling. I'm sorry about that situation. You'll have to forgive Joe asking you to do so much; we are just seriously limited in terms of actual developer time. Any little extra bit of information Joe can extract about the bug before a dev looks at it is invaluable. That does mean at times the user ends up doing more work than might be necessary, sadly.

**Comment by BlueDrink9** [comment]
In this case, I replicated it immediately on my install, so you can rest easy that it isn't just you.

It was a pretty easy sequence of steps to reproduce, so it would have been nice of Joe to do so if he had beta5 installed, but Joe already gives plenty of time to us so I don't begrudge it.

**Comment by BlueDrink9** [comment]
It's qt-only for me. GTK works fine

**Comment by BlueDrink9** [comment]
Bisected this problem to commit 9187259c68d431b98dc28664332a8dd613bc5911 Fix clear button not removing hotkey (#550)

**Comment by BlueDrink9** [comment]
Fixed in 20449de. My sincere apologies, this was a typo at my end. I must not have tested my fix on qt as well as gtk. I'll close this, but I will still look into the issues you have had reproducing the install.

**Comment by kreezxil** [comment]
thankyou, which beta do i install to get this? the same one? 0.96.5-beta?

**Comment by kreezxil** [comment]
I see you put out an update, so i reinstalled it, everything i've reported still crashes my copy of autokey. I wouldn't call this as closed, please reopen it.

**Comment by BlueDrink9** [comment]
I haven't released this fix as a beta yet. If you want to install it, you'll need to install a git version.

Edit: changed my mind, this is a serious enough issue to be worth a new beta. I'll make one asap

**Comment by BlueDrink9** [comment]
If you don't want to do that, you'll have to wait for more fixes to accumulate for a new beta

**Comment by kreezxil** [comment]
oh, ok, which git branch is it on.

**Comment by BlueDrink9** [comment]
It's on the `beta` branch, I just hadn't created the appropriate tag. But now you should be able to get it via `pip3 install --user https://github.com/autokey/autokey/archive/v0.96.0-beta.6.zip`. Thanks for your patience.

**Comment by kreezxil** [comment]
@BlueDrink9 should be noted, that installing the above mentioned beta 6, the autokey reports that b5 was installed.

```
Collecting https://github.com/autokey/autokey/archive/v0.96.0-beta.6.zip
  Downloading https://github.com/autokey/autokey/archive/v0.96.0-beta.6.zip
     / 463 kB 830 kB/s
Requirement already satisfied: pyinotify in /usr/lib/python3.9/site-packages (from autokey==0.96.0b5) (0.9.6)
Requirement already satisfied: python-xlib in /usr/lib/python3.9/site-packages (from autokey==0.96.0b5) (0.31)
Requirement already satisfied: packaging in /usr/lib/python3.9/site-packages (from autokey==0.96.0b5) (20.9)
Requirement already satisfied: pyparsing>=2.0.2 in /usr/lib/python3.9/site-packages (from packaging->autokey==0.96.0b5) (2.4.7)
Requirement already satisfied: six>=1.10.0 in /usr/lib/python3.9/site-packages (from python-xlib->autokey==0.96.0b5) (1.16.0)
Building wheels for collected packages: autokey
  Building wheel for autokey (setup.py) ... done
  Created wheel for autokey: filename=autokey-0.96.0b5-py3-none-any.whl size=394354 sha256=8a2a45c67408b6ba4997e2256f00e4b3a120f98f8ab44056b188f4627bc7980f
  Stored in directory: /root/.cache/pip/wheels/86/25/8b/a407ada37156e9de752a83c115660512f9eb8eaa8dfe2c6a98
Successfully built autokey
Installing collected packages: autokey
Successfully installed autokey-0.96.0b5
```

**Comment by BlueDrink9** [comment]
Known issue, but thanks for pointing it out. The only way to get the new release was to create it manually on github, which doesn't update the python metadata. The code is updated, but beta 6 won't show the correct version number in menus or anything

**Comment by kreezxil** [comment]
Renaming is working now! Thanks. You can close this now. :)
